File: app/src/test/java/com/coinninja/coinkeeper/receiver/WalletRegistrationCompleteReceiverTest.java
package com.coinninja.coinkeeper.receiver;

import android.content.Intent;

import com.coinninja.coinkeeper.TestCoinKeeperApplication;
import com.coinninja.coinkeeper.cn.service.CNGlobalMessagingService;
import com.coinninja.coinkeeper.service.DeviceRegistrationService;
import com.coinninja.coinkeeper.util.android.app.JobIntentService.JobServiceScheduler;

import org.hamcrest.Matchers;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class WalletRegistrationCompleteReceiverTest {

    private WalletRegistrationCompleteReceiver walletRegistrationCompleteReceiver;

    private TestCoinKeeperApplication application;

    @Before
    public void setUp() {
        application = (TestCoinKeeperApplication) RuntimeEnvironment.application;
        walletRegistrationCompleteReceiver = new WalletRegistrationCompleteReceiver();
    }

    @After
    public void tearDown() {
        application = null;
        walletRegistrationCompleteReceiver = null;
    }

    @Test
    public void schedules_device_registration_job_when_wallet_registration_has_completeds() {
        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);


        walletRegistrationCompleteReceiver.onReceive(application, null);

        verify(application.jobServiceScheduler).enqueueWork(eq(application),
                eq(DeviceRegistrationService.class),
                eq(101), captor.capture());
        Intent intent = captor.getValue();
        Assert.assertThat(intent.getComponent().getClassName(), Matchers.equalTo(DeviceRegistrationService.class.getName()));
    }

    @Test
    public void starts_global_messaging_services() {
        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);

        walletRegistrationCompleteReceiver.onReceive(application, null);

        verify(application.jobServiceScheduler).enqueueWork(eq(application),
                eq(CNGlobalMessagingService.class),
                eq(JobServiceScheduler.GLOBAL_MESSAGING_SERVICE_JOB_ID), captor.capture());
        Intent intent = captor.getValue();
        assertThat(intent.getComponent().getClassName(), equalTo(CNGlobalMessagingService.class.getName()));
    }

}

File: app/src/test/java/com/coinninja/coinkeeper/receiver/WalletRegistrationCompleteReceiverTest.kt
package com.coinninja.coinkeeper.receiver

import android.content.Intent
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.coinkeeper.TestCoinKeeperApplication
import com.coinninja.coinkeeper.cn.service.CNGlobalMessagingService
import com.coinninja.coinkeeper.util.android.app.JobIntentService.JobServiceScheduler
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.verify
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import androidx.test.ext.truth.content.IntentSubject.assertThat as assertThatIntent

@RunWith(AndroidJUnit4::class)
class WalletRegistrationCompleteReceiverTest {

    private val application: TestCoinKeeperApplication = ApplicationProvider.getApplicationContext()

    private fun createReceiver(): WalletRegistrationCompleteReceiver {
        val receiver = WalletRegistrationCompleteReceiver()
        receiver.onReceive(application, Intent())
        return receiver
    }

    @Test
    fun starts_global_messaging_services() {
        val captor: ArgumentCaptor<Intent> = ArgumentCaptor.forClass(Intent::class.java)
        val receiver = createReceiver()

        verify(receiver.jobServiceScheduler).enqueueWork(
                eq(application),
                eq(CNGlobalMessagingService::class.java),
                eq(JobServiceScheduler.GLOBAL_MESSAGING_SERVICE_JOB_ID),
                captor.capture())

        val intent = captor.value

        assertThatIntent(intent).hasComponent(application.packageName, CNGlobalMessagingService::class.java.name)
    }
}

